File: tools/stark-v-trace-dump/src/execute.rs
use super::decode::{DecodedInst, Opcode};
use super::elf::Memory;

// ---------------------------------------------------------------------------
// CPU state and executor
// ---------------------------------------------------------------------------

pub(super) struct Cpu {
    pub(super) regs: [u32; 32],
    pub(super) pc: u32,
}

impl Cpu {
    pub(super) fn new(entry_point: u32, stack_pointer: u32) -> Self {
        let mut regs = [0u32; 32];
        regs[2] = stack_pointer; // x2 = sp
        Self {
            regs,
            pc: entry_point,
        }
    }

    pub(super) fn read_reg(&self, r: u8) -> u32 {
        if r == 0 {
            0
        } else {
            self.regs[r as usize]
        }
    }

    fn write_reg(&mut self, r: u8, val: u32) {
        if r != 0 {
            self.regs[r as usize] = val;
        }
    }
}

// ---------------------------------------------------------------------------
// Executor
// ---------------------------------------------------------------------------

pub(super) enum ExecResult {
    Continue,
    Halt,
}

pub(super) fn execute(cpu: &mut Cpu, mem: &mut Memory, inst: &DecodedInst) -> ExecResult {
    let rs1_v = cpu.read_reg(inst.rs1);
    let rs2_v = cpu.read_reg(inst.rs2);

    match inst.opcode {
        // -- R-type arithmetic --
        Opcode::ADD => cpu.write_reg(inst.rd, rs1_v.wrapping_add(rs2_v)),
        Opcode::SUB => cpu.write_reg(inst.rd, rs1_v.wrapping_sub(rs2_v)),
        Opcode::XOR => cpu.write_reg(inst.rd, rs1_v ^ rs2_v),
        Opcode::OR => cpu.write_reg(inst.rd, rs1_v | rs2_v),
        Opcode::AND => cpu.write_reg(inst.rd, rs1_v & rs2_v),
        Opcode::SLL => cpu.write_reg(inst.rd, rs1_v << (rs2_v & 0x1F)),
        Opcode::SRL => cpu.write_reg(inst.rd, rs1_v >> (rs2_v & 0x1F)),
        Opcode::SRA => {
            let signed = rs1_v as i32;
            cpu.write_reg(inst.rd, (signed >> (rs2_v & 0x1F)) as u32);
        }
        Opcode::SLT => {
            let a = rs1_v as i32;
            let b = rs2_v as i32;
            cpu.write_reg(inst.rd, if a < b { 1 } else { 0 });
        }
        Opcode::SLTU => {
            cpu.write_reg(inst.rd, if rs1_v < rs2_v { 1 } else { 0 });
        }

        // -- I-type arithmetic --
        Opcode::ADDI => cpu.write_reg(inst.rd, rs1_v.wrapping_add(inst.imm as u32)),
        Opcode::XORI => cpu.write_reg(inst.rd, rs1_v ^ (inst.imm as u32)),
        Opcode::ORI => cpu.write_reg(inst.rd, rs1_v | (inst.imm as u32)),
        Opcode::ANDI => cpu.write_reg(inst.rd, rs1_v & (inst.imm as u32)),
        Opcode::SLLI => cpu.write_reg(inst.rd, rs1_v << (inst.imm as u32 & 0x1F)),
        Opcode::SRLI => cpu.write_reg(inst.rd, rs1_v >> (inst.imm as u32 & 0x1F)),
        Opcode::SRAI => {
            let signed = rs1_v as i32;
            cpu.write_reg(inst.rd, (signed >> (inst.imm as u32 & 0x1F)) as u32);
        }
        Opcode::SLTI => {
            let a = rs1_v as i32;
            cpu.write_reg(inst.rd, if a < inst.imm { 1 } else { 0 });
        }
        Opcode::SLTIU => {
            cpu.write_reg(inst.rd, if rs1_v < (inst.imm as u32) { 1 } else { 0 });
        }

        // -- Loads --
        Opcode::LB => {
            let addr = rs1_v.wrapping_add(inst.imm as u32);
            let byte = mem.read_byte(addr);
            let signed = byte as i8;
            cpu.write_reg(inst.rd, signed as i32 as u32);
        }
        Opcode::LBU => {
            let addr = rs1_v.wrapping_add(inst.imm as u32);
            cpu.write_reg(inst.rd, mem.read_byte(addr) as u32);
        }
        Opcode::LH => {
            let addr = rs1_v.wrapping_add(inst.imm as u32);
            let half = mem.read_u16(addr);
            let signed = half as i16;
            cpu.write_reg(inst.rd, signed as i32 as u32);
        }
        Opcode::LHU => {
            let addr = rs1_v.wrapping_add(inst.imm as u32);
            cpu.write_reg(inst.rd, mem.read_u16(addr) as u32);
        }
        Opcode::LW => {
            let addr = rs1_v.wrapping_add(inst.imm as u32);
            cpu.write_reg(inst.rd, mem.read_u32(addr));
        }

        // -- Stores --
        Opcode::SB => {
            let addr = rs1_v.wrapping_add(inst.imm as u32);
            mem.write_byte(addr, rs2_v as u8);
        }
        Opcode::SH => {
            let addr = rs1_v.wrapping_add(inst.imm as u32);
            mem.write_u16(addr, rs2_v as u16);
        }
        Opcode::SW => {
            let addr = rs1_v.wrapping_add(inst.imm as u32);
            mem.write_u32(addr, rs2_v);
        }

        // -- Branches --
        Opcode::BEQ => {
            if rs1_v == rs2_v {
                cpu.pc = cpu.pc.wrapping_add(inst.imm as u32);
                return ExecResult::Continue;
            }
        }
        Opcode::BNE => {
            if rs1_v != rs2_v {
                cpu.pc = cpu.pc.wrapping_add(inst.imm as u32);
                return ExecResult::Continue;
            }
        }
        Opcode::BLT => {
            if (rs1_v as i32) < (rs2_v as i32) {
                cpu.pc = cpu.pc.wrapping_add(inst.imm as u32);
                return ExecResult::Continue;
            }
        }
        Opcode::BGE => {
            if (rs1_v as i32) >= (rs2_v as i32) {
                cpu.pc = cpu.pc.wrapping_add(inst.imm as u32);
                return ExecResult::Continue;
            }
        }
        Opcode::BLTU => {
            if rs1_v < rs2_v {
                cpu.pc = cpu.pc.wrapping_add(inst.imm as u32);
                return ExecResult::Continue;
            }
        }
        Opcode::BGEU => {
            if rs1_v >= rs2_v {
                cpu.pc = cpu.pc.wrapping_add(inst.imm as u32);
                return ExecResult::Continue;
            }
        }

        // -- Jumps --
        Opcode::JAL => {
            cpu.write_reg(inst.rd, cpu.pc.wrapping_add(4));
            cpu.pc = cpu.pc.wrapping_add(inst.imm as u32);
            return ExecResult::Continue;
        }
        Opcode::JALR => {
            let target = (rs1_v.wrapping_add(inst.imm as u32)) & 0xFFFF_FFFE;
            cpu.write_reg(inst.rd, cpu.pc.wrapping_add(4));
            cpu.pc = target;
            return ExecResult::Continue;
        }

        // -- Upper immediates --
        Opcode::LUI => cpu.write_reg(inst.rd, inst.imm as u32),
        Opcode::AUIPC => cpu.write_reg(inst.rd, cpu.pc.wrapping_add(inst.imm as u32)),

        // -- RV32M: Multiply --
        Opcode::MUL => {
            let result = (rs1_v as i32 as i64).wrapping_mul(rs2_v as i32 as i64) as u64;
            cpu.write_reg(inst.rd, result as u32);
        }
        Opcode::MULH => {
            let a = rs1_v as i32 as i64;
            let b = rs2_v as i32 as i64;
            let product = a.wrapping_mul(b) as u64;
            cpu.write_reg(inst.rd, (product >> 32) as u32);
        }
        Opcode::MULHSU => {
            let a = rs1_v as i32 as i64;
            let b = rs2_v as u64 as i64;
            let product = a.wrapping_mul(b) as u64;
            cpu.write_reg(inst.rd, (product >> 32) as u32);
        }
        Opcode::MULHU => {
            let a = rs1_v as u64;
            let b = rs2_v as u64;
            let product = a.wrapping_mul(b);
            cpu.write_reg(inst.rd, (product >> 32) as u32);
        }

        // -- RV32M: Divide --
        Opcode::DIV => {
            let a = rs1_v as i32;
            let b = rs2_v as i32;
            if b == 0 {
                cpu.write_reg(inst.rd, (-1i32) as u32);
            } else if a == i32::MIN && b == -1 {
                cpu.write_reg(inst.rd, a as u32);
            } else {
                cpu.write_reg(inst.rd, a.wrapping_div(b) as u32);
            }
        }
        Opcode::DIVU => {
            let a = rs1_v;
            let b = rs2_v;
            if b == 0 {
                cpu.write_reg(inst.rd, 0xFFFF_FFFF);
            } else {
                cpu.write_reg(inst.rd, a / b);
            }
        }
        Opcode::REM => {
            let a = rs1_v as i32;
            let b = rs2_v as i32;
            if b == 0 {
                cpu.write_reg(inst.rd, a as u32);
            } else if a == i32::MIN && b == -1 {
                cpu.write_reg(inst.rd, 0);
            } else {
                cpu.write_reg(inst.rd, a.wrapping_rem(b) as u32);
            }
        }
        Opcode::REMU => {
            let a = rs1_v;
            let b = rs2_v;
            if b == 0 {
                cpu.write_reg(inst.rd, a);
            } else {
                cpu.write_reg(inst.rd, a % b);
            }
        }

        // -- System --
        Opcode::ECALL => return ExecResult::Halt,
        Opcode::EBREAK => return ExecResult::Halt,
    }

    // Default: advance PC by 4 (branches/jumps return early).
    cpu.pc = cpu.pc.wrapping_add(4);
    ExecResult::Continue
}

// ---------------------------------------------------------------------------
// Main runner
// ---------------------------------------------------------------------------

File: src/frontends/riscv/runner/execute.zig
//! Fail-closed legacy execution boundary.
//!
//! Every proof-bearing RV32IM opcode retires through `generated_retirement`;
//! this module retains only the host-visible system-instruction errors. Keeping
//! the old entry point explicit prevents a second mutable semantics authority
//! from silently reappearing.

const std = @import("std");
const Cpu = @import("cpu.zig").Cpu;
const Memory = @import("memory.zig").Memory;
const decode = @import("decode.zig");
const Opcode = decode.Opcode;
const DecodedInst = decode.DecodedInst;

pub const ExecuteError = error{
    Ecall,
    Ebreak,
    /// This opcode has migrated to a generated, failure-atomic retirement
    /// authority and must be dispatched before entering the legacy executor.
    GeneratedRetirementRequired,
    MisalignedMemoryAccess,
    InstructionAddressMisaligned,
};

/// Execute a single decoded instruction, mutating `cpu` and `mem`.
/// Returns system-instruction errors to the host and rejects every opcode now
/// owned by generated failure-atomic retirement.
pub fn execute(cpu: *Cpu, mem: *Memory, inst: DecodedInst) ExecuteError!void {
    _ = mem;
    switch (inst.opcode) {
        // ----------------------------------------------------------------
        // R-type arithmetic
        // ----------------------------------------------------------------
        .ADD, .SUB, .XOR, .OR, .AND => return error.GeneratedRetirementRequired,
        .SLL, .SRL, .SRA, .SLT, .SLTU => return error.GeneratedRetirementRequired,

        // ----------------------------------------------------------------
        // I-type arithmetic
        // ----------------------------------------------------------------
        .ADDI, .XORI, .ORI, .ANDI => return error.GeneratedRetirementRequired,
        .SLLI, .SRLI, .SRAI => return error.GeneratedRetirementRequired,
        .SLTI, .SLTIU => return error.GeneratedRetirementRequired,

        // ----------------------------------------------------------------
        // Loads (I-type)
        // ----------------------------------------------------------------
        .LB, .LBU, .LH, .LHU, .LW => return error.GeneratedRetirementRequired,

        // ----------------------------------------------------------------
        // Stores (S-type)
        // ----------------------------------------------------------------
        .SB, .SH, .SW => return error.GeneratedRetirementRequired,

        // ----------------------------------------------------------------
        // Branches (B-type)
        // ----------------------------------------------------------------
        .BEQ, .BNE => return error.GeneratedRetirementRequired,
        .BLT, .BLTU, .BGE, .BGEU => return error.GeneratedRetirementRequired,

        // ----------------------------------------------------------------
        // Jumps
        // ----------------------------------------------------------------
        .JAL => return error.GeneratedRetirementRequired,
        .JALR => return error.GeneratedRetirementRequired,

        // ----------------------------------------------------------------
        // Upper immediates
        // ----------------------------------------------------------------
        .LUI => return error.GeneratedRetirementRequired,
        .AUIPC => return error.GeneratedRetirementRequired,

        // ----------------------------------------------------------------
        // RV32M: Multiply / Divide
        // ----------------------------------------------------------------
        .MUL,
        .MULH,
        .MULHSU,
        .MULHU,
        .DIV,
        .DIVU,
        .REM,
        .REMU,
        => return error.GeneratedRetirementRequired,

        // ----------------------------------------------------------------
        // System
        // ----------------------------------------------------------------
        .FENCE => return error.GeneratedRetirementRequired,
        .ECALL => return error.Ecall,
        .EBREAK => return error.Ebreak,
    }

    // Default: advance PC by 4 (branches/jumps return early).
    cpu.pc +%= 4;
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

fn makeTestCpuAndMem() struct { cpu: Cpu, mem: Memory } {
    return .{
        .cpu = Cpu.init(0x1000, 0x8000_0000),
        .mem = Memory.init(std.testing.allocator),
    };
}

test "legacy execute refuses production-migrated BASE_ALU_REG" {
    var t = makeTestCpuAndMem();
    defer t.mem.deinit();
    t.cpu.writeReg(2, 10);
    t.cpu.writeReg(3, 20);
    const words = [_]u32{
        0x003100B3, // ADD x1, x2, x3
        0x403100B3, // SUB x1, x2, x3
        0x003140B3, // XOR x1, x2, x3
        0x003160B3, // OR  x1, x2, x3
        0x003170B3, // AND x1, x2, x3
    };
    for (words) |word| {
        const before = t.cpu;
        try std.testing.expectError(
            error.GeneratedRetirementRequired,
            execute(&t.cpu, &t.mem, try DecodedInst.decode(word)),
        );
        try std.testing.expectEqualDeep(before, t.cpu);
    }
}

test "legacy execute refuses production-migrated LT_REG and register shifts" {
    inline for ([_]Opcode{ .SLT, .SLTU, .SLL, .SRL, .SRA }) |opcode| {
        var t = makeTestCpuAndMem();
        defer t.mem.deinit();
        t.cpu.writeReg(2, 0x8000_0001);
        t.cpu.writeReg(3, 7);
        const before = t.cpu;
        try std.testing.expectError(
            error.GeneratedRetirementRequired,
            execute(&t.cpu, &t.mem, .{
                .opcode = opcode,
                .rd = 1,
                .rs1 = 2,
                .rs2 = 3,
                .imm = 0,
            }),
        );
        try std.testing.expectEqualDeep(before, t.cpu);
    }
}

test "legacy execute refuses production-migrated BASE_ALU_IMM" {
    var t = makeTestCpuAndMem();
    defer t.mem.deinit();
    t.cpu.writeReg(1, 100);
    const words = [_]u32{
        0x02A08293, // ADDI x5, x1, 42
        0x02A0C293, // XORI x5, x1, 42
        0x02A0E293, // ORI x5, x1, 42
        0x02A0F293, // ANDI x5, x1, 42
    };
    for (words) |word| {
        const before = t.cpu;
        try std.testing.expectError(
            error.GeneratedRetirementRequired,
            execute(&t.cpu, &t.mem, try DecodedInst.decode(word)),
        );
        try std.testing.expectEqualDeep(before, t.cpu);
    }
}

test "legacy execute refuses production-migrated immediate shifts" {
    inline for ([_]Opcode{ .SLLI, .SRLI, .SRAI }) |opcode| {
        var t = makeTestCpuAndMem();
        defer t.mem.deinit();
        t.cpu.writeReg(2, 0x8000_0001);
        const before = t.cpu;
        try std.testing.expectError(
            error.GeneratedRetirementRequired,
            execute(&t.cpu, &t.mem, .{
                .opcode = opcode,
                .rd = 1,
                .rs1 = 2,
                .rs2 = 0,
                .imm = 7,
            }),
        );
        try std.testing.expectEqualDeep(before, t.cpu);
    }
}

test "legacy execute refuses production-migrated LOAD_STORE without mutation" {
    inline for ([_]Opcode{ .LB, .LH, .LW, .LBU, .LHU, .SB, .SH, .SW }) |opcode| {
        var t = makeTestCpuAndMem();
        defer t.mem.deinit();
        t.cpu.writeReg(2, 0x2000);
        t.cpu.writeReg(3, 0xCAFE_BABE);
        t.mem.writeU32(0x2000, 0x1234_5678);
        const inst = DecodedInst{
            .opcode = opcode,
            .rd = 1,
            .rs1 = 2,
            .rs2 = 3,
            .imm = 0,
        };
        const before = t.cpu;
        try std.testing.expectError(
            error.GeneratedRetirementRequired,
            execute(&t.cpu, &t.mem, inst),
        );
        try std.testing.expectEqualDeep(before, t.cpu);
        try std.testing.expectEqual(@as(u32, 0x1234_5678), t.mem.readU32(0x2000));
    }
}

test "legacy execute refuses production-migrated BRANCH_EQ opcodes" {
    inline for ([_]u32{
        0x0020_8463, // BEQ x1, x2, +8
        0x0020_9463, // BNE x1, x2, +8
    }) |word| {
        var t = makeTestCpuAndMem();
        defer t.mem.deinit();
        t.cpu.writeReg(1, 42);
        t.cpu.writeReg(2, 42);
        const inst = try DecodedInst.decode(word);
        const before = t.cpu;
        try std.testing.expectError(
            error.GeneratedRetirementRequired,
            execute(&t.cpu, &t.mem, inst),
        );
        try std.testing.expectEqualDeep(before, t.cpu);
    }
}

test "legacy execute refuses production-migrated BRANCH_LT opcodes" {
    inline for ([_]u32{
        0x0062_c463, // BLT  x5, x6, +8
        0x0062_e463, // BLTU x5, x6, +8
        0x0053_5463, // BGE  x6, x5, +8
        0x0053_7463, // BGEU x6, x5, +8
    }) |word| {
        var t = makeTestCpuAndMem();
        defer t.mem.deinit();
        t.cpu.writeReg(5, 0x8000_0000);
        t.cpu.writeReg(6, 0xffff_ffff);
        const inst = try DecodedInst.decode(word);
        const before = t.cpu;
        try std.testing.expectError(
            error.GeneratedRetirementRequired,
            execute(&t.cpu, &t.mem, inst),
        );
        try std.testing.expectEqualDeep(before, t.cpu);
    }
}

test "legacy execute refuses production-migrated LT_IMM opcodes" {
    inline for ([_]u32{
        0x0002_a313, // SLTI  x6, x5, 0
        0xfff2_b393, // SLTIU x7, x5, -1
    }) |word| {
        var t = makeTestCpuAndMem();
        defer t.mem.deinit();
        t.cpu.writeReg(5, 0x8000_0000);
        const inst = try DecodedInst.decode(word);
        const before = t.cpu;
        try std.testing.expectError(
            error.GeneratedRetirementRequired,
            execute(&t.cpu, &t.mem, inst),
        );
        try std.testing.expectEqualDeep(before, t.cpu);
    }
}

test "legacy execute refuses production-migrated JAL" {
    var t = makeTestCpuAndMem();
    defer t.mem.deinit();
    // JAL x1, +0 => 0x000000EF
    const inst = try DecodedInst.decode(0x000000EF);
    const before = t.cpu;
    try std.testing.expectError(
        error.GeneratedRetirementRequired,
        execute(&t.cpu, &t.mem, inst),
    );
    try std.testing.expectEqualDeep(before, t.cpu);
}

test "legacy execute refuses production-migrated JALR" {
    var t = makeTestCpuAndMem();
    defer t.mem.deinit();
    t.cpu.writeReg(2, 0x1000);
    const inst = try DecodedInst.decode(0x0041_00e7); // JALR x1, x2, +4
    const before = t.cpu;
    try std.testing.expectError(
        error.GeneratedRetirementRequired,
        execute(&t.cpu, &t.mem, inst),
    );
    try std.testing.expectEqualDeep(before, t.cpu);
}

test "legacy execute refuses production-migrated LUI" {
    var t = makeTestCpuAndMem();
    defer t.mem.deinit();
    // LUI x1, 0x12345 => 0x123450B7
    const inst = try DecodedInst.decode(0x123450B7);
    const before = t.cpu;
    try std.testing.expectError(
        error.GeneratedRetirementRequired,
        execute(&t.cpu, &t.mem, inst),
    );
    try std.testing.expectEqualDeep(before, t.cpu);
}

test "legacy execute refuses production-migrated AUIPC" {
    var t = makeTestCpuAndMem();
    defer t.mem.deinit();
    const inst = try DecodedInst.decode(0x1234_5097);
    const before = t.cpu;
    try std.testing.expectError(
        error.GeneratedRetirementRequired,
        execute(&t.cpu, &t.mem, inst),
    );
    try std.testing.expectEqualDeep(before, t.cpu);
}

test "legacy execute refuses production-migrated FENCE" {
    var t = makeTestCpuAndMem();
    defer t.mem.deinit();
    const inst = try DecodedInst.decode(0xf538_8f8f);
    const before = t.cpu;
    try std.testing.expectError(
        error.GeneratedRetirementRequired,
        execute(&t.cpu, &t.mem, inst),
    );
    try std.testing.expectEqualDeep(before, t.cpu);
}

test "legacy execute refuses production-migrated RV32M without mutation" {
    inline for ([_]Opcode{ .MUL, .MULH, .MULHSU, .MULHU, .DIV, .DIVU, .REM, .REMU }) |opcode| {
        var t = makeTestCpuAndMem();
        defer t.mem.deinit();
        t.cpu.writeReg(2, 100);
        t.cpu.writeReg(3, 7);
        const before = t.cpu;
        try std.testing.expectError(
            error.GeneratedRetirementRequired,
            execute(&t.cpu, &t.mem, .{
                .opcode = opcode,
                .rd = 1,
                .rs1 = 2,
                .rs2 = 3,
                .imm = 0,
            }),
        );
        try std.testing.expectEqualDeep(before, t.cpu);
    }
}

test "execute ECALL returns error" {
    var t = makeTestCpuAndMem();
    defer t.mem.deinit();
    // The runner synthesizes ECALL (SYSTEM words are outside the pinned
    // decode contract); execute keeps its syscall error surface.
    const inst = DecodedInst{ .opcode = .ECALL, .rd = 0, .rs1 = 0, .rs2 = 0, .imm = 0 };
    const result = execute(&t.cpu, &t.mem, inst);
    try std.testing.expectError(error.Ecall, result);
}
